API Remove bind_manipulation_capture RequestFilter in favour of Injection overrides

robbieaverill opened the pull request:
This pull request re-enables support for PostgreSQL and SQLite3 databases, as well as removing the RequestFilter (removes the need for #164) and switching to using the Injector to overload the database classes instead of PHP via the RequestFilter.

Locally my test suite slowed down about 400% with this change, but I don't see any change in Travis between this build and other, older builds.

There are a couple of implications that I discussed with @tractorcow on this.

The auditor module also does this, so we'll need to work out a way for that module to monitor database manipulations in the same way as this module, while letting all three run correctly.

The purpose of doing this in the first place is to capture all requests to `DB::manipulate`, which really only get called in `DataObject::onAfterWrite` and `ManyManyList::add`. It uses this information to trigger specific reindexing.

We could switch to using the `augmentWrite` extension point for DataObject in SS 4.0.0, which would capture pretty much all of the same data, provided we ensure that fulltextsearch extensions are added last. This would also work for the auditor module. If we did this we'd need to add an extension point to the ManyManyList to support the same behaviour, which probably wouldn't make it until SS 4.1.0.

We could also add onBeforeManipulate and onAfterManipulate hooks to the `DB` class, which would need to be made Extensible as well.

By doing this we would ensure that other modules could do the same types of things as this without having to worry about an inheritance chain. It would also mean that we could remove the three SearchManipulateCapture_* classes from this module too.

I'd like to get this in for now, and look at an improvement for this in the near future.

Resolves #161 and resolves #160 

---

Also, the `testSoftCap` unit test fails now with this change when run in combination with the rest of the test class, which I assume is a problem with the global state. It passes when run in isolation. For now I don't have a lot of time to debug it (I time boxed an hour) so I've marked it as skipped with a todo to fix it.

dhensby commented:
@robbieaverill could be failing because `setUpOnce` is not a valid API anymore. Use `setUpBeforeClass`

if that doesn't fix it, where in the test is it failing?

dhensby commented:
Alternatively add this to the end of your `setUp`:

```php
        // this is needed because the config locator cache overrides changes to injector config
        Injector::inst()->setConfigLocator(new SilverStripeServiceConfigurationLocator());
```

I've found I've had to do this when dealing with injector/config changes

robbieaverill commented:
@dhensby I'm not exactly sure why, but yeah - changing the setUpOnce to setUpBeforeClass fixes it. Reason I say I'm not sure why is because the initial `$illegalExtensions` declaration hadn't been updated yet anyway, so fixing setUpOnce shouldn't have had any impact.

Nevertheless tests are all passing now - cheers

robbieaverill commented:
@dhensby yeah so I've updated setUpOnce etc, but it wouldn't have been effective anyway since the `$illegalExtensions` hadn't been updated. I've updated both now anyway.

The problem is in Postgres only:

```
$ vendor/bin/phpunit --filter BatchedProcessorTest vendor/silverstripe/fulltextsearch/tests '' flush=1
PHPUnit 5.7.23 by Sebastian Bergmann and contributors.

..F                                                                 3 / 3 (100%)

Time: 17.9 seconds, Memory: 80.00MB

There was 1 failure:

1) SilverStripe\FullTextSearch\Tests\BatchedProcessorTest::testSoftCap
Failed asserting that 17 matches expected 8.

/Users/robbieaverill/dev/ss4/vendor/silverstripe/fulltextsearch/tests/BatchedProcessorTest.php:227
/Users/robbieaverill/dev/ss4/vendor/phpunit/phpunit/phpunit:52

FAILURES!
Tests: 3, Assertions: 38, Failures: 1.
```

Which I assume is a part of the global state of the indexer or processor class leaking between tests, but I wasn't able to track it down.

robbieaverill commented:
I've done a bit of work in #187 to make the test state a bit more manageable, so will try this again with that in place to see if it's helped at all.

robbieaverill commented:
Nope, still failing on Postgres. If you're happy to leave the test skipped for now @dhensby that'd be great.

NightJar commented on src/Solr/Reindex/Jobs/SolrReindexQueuedJobBase.php:
```diff
@@ -138,6 +138,11 @@ public function getJobType()
         return QueuedJob::QUEUED;
     }
 
+    public function getSignature()
```
Not important to this review since it's pre-existing, but I feel like this should be named something more like `generateSignature`, since it's not fetching anything, and would return a different value every time, even when called on the same object.

robbieaverill replied:
Yeah, this is a queuedjobs API unfortunately

NightJar replied:
oic, welp, that's fine then :)

NightJar commented on tests/BatchedProcessorTest.php:
```diff
@@ -196,21 +200,24 @@ public function testMultipleCommits()
      */
     public function testSoftCap()
     {
+        $this->markTestSkipped(
```
With the accompanied comment I feel like `markTestIncomplete` would probably be more ideal.

robbieaverill commented on src/Search/Processors/SearchUpdateCommitJobProcessor.php:
```diff
@@ -22,9 +22,9 @@ class SearchUpdateCommitJobProcessor implements QueuedJob
      *
      * @config
-     * @var int
+     * @var string
```
They're ints but cast as strings in the QueuedJob interface

tractorcow replied:
`int`eresting.

NightJar replied:
Whoops, truth.

NightJar commented on src/Search/Processors/SearchUpdateQueuedJobProcessor.php:
```diff
@@ -16,9 +16,9 @@ class SearchUpdateQueuedJobProcessor extends SearchUpdateBatchedProcessor implem
      * The QueuedJob queue to use when processing updates
      * @config
-     * @var int
+     * @var string
```
Wrong.
